### PR TITLE
TEST-#5378: port stack, unstack, replace and groups benchmarks from pandas

### LINKED ISSUE #5378: [ASV] port stack, unstack, replace and  groups benchmarks from pandas github
Have to add ASV benchmarks related which does the following actions
1) perform stack and unstack on dataframe.
2) Replace function on a dataframe.
3) Get groups returned by groupby.

### PR DESCRIPTION
**Pull request by arunjose696**
## What do these changes do?

Adding 3 benchmark classes from pandas benchmark TimeSimpleReshape , TimeReplace and TimeGroups.

Source PRs #5055 , #5056 , #5057 

- [x] first commit message and PR title follow format outlined [here](https://modin.readthedocs.io/en/latest/development/contributing.html#commit-message-formatting)
  > **_NOTE:_**  If you edit the PR title to match this format, you need to add another commit (even if it's empty) or amend your last commit for the CI job that checks the PR title to pick up the new PR title.
- [x] passes `flake8 modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] passes `black --check modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] signed commit with `git commit -s` 
- [x] Resolves #5378 
- [x] tests added and passing
- [x] module layout described at `docs/development/architecture.rst` is up-to-date

### CODE REVIEW

File: asv_bench/benchmarks/benchmarks.py [hunk 1]
```diff
@@ -1113,4 +1113,63 @@ def time_remove_categories(self, shape):
...
+
+    def setup(self, shape):
+        self.series = IMPL.Series(np.random.randint(0, 100, size=shape[0]))
+        execute(self.series)
+
+    # returns a pretty dict thus not calling execute
+    def time_series_groups(self, shape):
+        self.series.groupby(self.series).groups
+
+    # returns a  dict thus not calling execute
```
Comment (anmyachev) [inline asv_bench/benchmarks/benchmarks.py:1170]:
```suggestion
    # returns a dict thus not calling execute
```

File: asv_bench/benchmarks/benchmarks.py [hunk 2]
```diff
@@ -1113,4 +1113,63 @@ def time_remove_categories(self, shape):
...
+
+    params = [get_benchmark_shapes("TimeReplace")]
+    param_names = ["shape"]
+
+    def setup(self, shape):
+        rows, cols = shape
+        self.to_replace = {i: getattr(IMPL, "Timestamp")(i) for i in range(rows)}
+        self.df = IMPL.DataFrame()
+        for col in range(cols):
+            self.df["col" + str(col + 1)] = np.random.randint(rows, size=rows)
```
Comment (anmyachev) [inline asv_bench/benchmarks/benchmarks.py:1150]:
```suggestion
        columns = [f"col{col + 1}" for col in range(cols)]
        self.df = IMPL.DataFrame(np.random.randint(rows, size=(rows, cols), columns=columns)
```

**Review by anmyachev — CHANGES_REQUESTED** [review]

File: asv_bench/benchmarks/benchmarks.py [hunk 3]
```diff
@@ -1113,4 +1113,63 @@ def time_remove_categories(self, shape):
...
+        index = IMPL.MultiIndex.from_arrays(arrays)
+        self.df = IMPL.DataFrame(np.random.randn(rows, cols), index=index)
+        self.udf = self.df.unstack(1)
+        execute(self.df), execute(self.udf)
+
+    def time_stack(self, shape):
+        execute(self.udf.stack())
+
+    def time_unstack(self, shape):
+        execute(self.df.unstack(1))
```
Comment (anmyachev) [inline asv_bench/benchmarks/benchmarks.py:1137]:
```suggestion
class BaseReshape:
    def setup(self, shape):
        rows, cols = shape
        k = 10
        arrays = [
            np.arange(rows // k).repeat(k),
            np.roll(np.tile(np.arange(rows // k), k), 25),
        ]
        index = IMPL.MultiIndex.from_arrays(arrays)
        self.df = IMPL.DataFrame(np.random.randn(rows, cols), index=index)
        execute(self.df)


class TimeStack(BaseReshape):
    params = [get_benchmark_shapes("TimeStack")]
    param_names = ["shape"]

    def setup(self, shape):
        super().setup(shape)
        self.udf = self.df.unstack(1)
        execute(self.udf)

    def time_stack(self, shape):
        execute(self.udf.stack())

class TimeUnstack(BaseReshape):
    params = [get_benchmark_shapes("TimeUnstack")]
    param_names = ["shape"]

    def time_unstack(self, shape):
        execute(self.df.unstack(1))
```

File: asv_bench/benchmarks/utils/data_shapes.py [hunk 4]
```diff
@@ -101,6 +101,8 @@
             "TimeReindexMethod",
             "TimeFillnaMethodDataframe",
             "TimeDropDuplicatesDataframe",
+            "TimeSimpleReshape",
```
Comment (anmyachev) [inline asv_bench/benchmarks/utils/data_shapes.py:104]:
```suggestion
            "TimeStack",
            "TimeUnstack",
```

File: asv_bench/benchmarks/benchmarks.py (hunk 2 above)
Comment (anmyachev) [reply asv_bench/benchmarks/benchmarks.py:1150]: Creating a dataframe in Modin by assigning columns in a loop results in inefficient internal partitioning, which in turn leads to poor performance.

Comment (arunjose696) [reply asv_bench/benchmarks/benchmarks.py]: changed this to normal initialization of dataframe.

**Review by anmyachev — DISMISSED** [review]
LGTM!

**Review by anmyachev — APPROVED** [review]
LGTM! Thanks @arunjose696

**Pull request merged**